Rework cache timestamps

keszybz opened the pull request.

bluca reviewed (approved):
LGTM! Very nice improvements - when testing this functionality, I had systems with the clock starting way in the past, but none with the clock in the future, so I missed this corner case. Thanks for fixing it!

poettering commented on src/core/manager.c:
```diff
@@ -1956,9 +1956,12 @@ bool manager_unit_file_maybe_loadable_from_cache(Unit *u) {
         if (u->manager->unit_cache_mtime == 0)
```
can you add a comment here, too, explaining what precisely this checks?

keszybz replied:
Hmm, it was added by @bluca  in d904afc730268d50502f764dfd55b8cf4906c46f. I didn't think too much about it. But looking at it carefully, I'm not sure if I understand the purpose. @bluca, can you explain? If we remove it, we would be try to reload the unit if if the cache is not initialized. I'm not sure if that could ever happen (because we load the cache quite early), but either way, I don't see a point for the check.

bluca replied:
If memory serves, there was another pre-existing check somewhere else that skipped using _mtime if it was not initialised, so I replicated it. I don't remember any specific reason apart from this, so I think it can be removed.

keszybz replied:
Oh, I think I see: unit_file_build_name_map() had a check like this, but it is used in the *inverse* sense: we always rebuild the cache if the timestamp == 0. So yeah, it seems appropriate to remove this.

poettering commented on src/shared/unit-file.c:
```diff
@@ -217,18 +224,20 @@ bool lookup_paths_mtime_good(const LookupPaths *lp, usec_t mtime) {
         }
 
-        return true;
+        uint64_t new = siphash24_finalize(&state);
```
this compiles? `new()` is a macro for us...

(I guess it compiles for the only reason that there's no `(` following right after.)

Please rename this variable

poettering commented on src/shared/unit-file.c:
```diff
@@ -217,18 +224,20 @@ bool lookup_paths_mtime_good(const LookupPaths *lp, usec_t mtime) {
+        if (ret_new)
+                *ret_new = new;
+        if (new != timestamp)
+                log_debug_errno(errno, "Modification times have changed, need to update cache.");
```
`errno` has no relevant value here, please use plain log_debug()

poettering commented on src/core/manager.h:
```diff
@@ -233,7 +233,7 @@ struct Manager {
         Hashmap *unit_name_map;
         Set *unit_path_cache;
-        usec_t unit_cache_mtime;
+        uint64_t unit_cache_timestamp;
```
hmm, "timestamp" is not the best name if this is not actually a timestamp. Maybe `unit_cache_hashed_timestamp`,  or so would be a better name?

keszybz replied:
I went with `_timestamp_hash`.

poettering reviewed (changes requested):
looks good, some minor comments

poettering commented:
hmm, i wonder if a simple hash like XOR might be more appropriate here. i.e. is it a good thing to be dependent of search path order or a bad? Might be it's good, dunno, just wanted to raise the question? opinion?

keszybz commented:
> hmm, i wonder if a simple hash like XOR might be more appropriate here. i.e. is it a good thing to be dependent of search path order or a bad? Might be it's good, dunno, just wanted to raise the question? opinion?

The search path order is fixed, so it doesn't really matter. I don't have a very strong opinion on siphash vs. xor: we're only hashing on the order of 10 values, and the hash cost is swamped by IO. But we don't need a strong hash. I can rework to use something else if you think it'd be better.

keszybz commented:
Updated. I added one more separate commit to drop the check you asked about, with explanation in the commit message. I also added one more comment in that function. Otherwise, the only change is to rename the variables.

anitazha commented:
bionic-i386 timed out like usual. merging
